src/java/org/apache/cassandra/utils/NoWaitRateLimiter.java

```diff
@@ -0,0 +1,261 @@
+    {
+        return create(permitsPerSecond, burstSeconds, Ticker.systemTicker());
+    }
+
+    public static NoWaitRateLimiter create(double permitsPerSecond, double burstSeconds, Ticker ticker)
+    {
+        SmoothRateLimiter delegate = new SmoothRateLimiter(burstSeconds);
+        NoWaitRateLimiter limiter = new NoWaitRateLimiter(delegate, ticker);
+        limiter.setRate(permitsPerSecond);
+        return limiter;
+    }
+    
+    @VisibleForTesting
+    public synchronized void reset(double permitsPerSecond, double burstSeconds, Ticker ticker)
+    {
+        this.stopwatch = Stopwatch.createStarted(ticker);
+        this.delegate = new SmoothRateLimiter(burstSeconds);
+        setRate(permitsPerSecond);
+    }
+
+    public final void setRate(double permitsPerSecond) 
+    {
+        Preconditions.checkArgument(permitsPerSecond > 0.0D && !Double.isNaN(permitsPerSecond), "rate must be positive");
+        
+        synchronized (this) 
+        {
+            delegate.doSetRate(permitsPerSecond, stopwatch.elapsed(TimeUnit.MICROSECONDS));
+        }
+    }
+
+    public synchronized double getRate() 
+    {
+        return delegate.getRate();
+    }
+
+    /**
+     * Forces the acquisition of a single permit.
+     * 
+     * @return microseconds until the acquired permit is available, and zero if it already is
+     */
+    public synchronized long reserveAndGetWaitLength() 
+    {
+        long nowMicros = this.stopwatch.elapsed(TimeUnit.MICROSECONDS);
+        long momentAvailable = delegate.reserveEarliestAvailable(1, nowMicros);
+        return Math.max(momentAvailable - nowMicros, 0L);
+    }
+
+    public synchronized boolean tryAcquire()
+    {
+        long nowMicros = this.stopwatch.elapsed(TimeUnit.MICROSECONDS);
+
+        if (!delegate.canAcquire(nowMicros)) 
+        {
+            return false;
+        }
+
+        delegate.reserveEarliestAvailable(1, nowMicros);
+            
+        return true;
+    }
+
+    public synchronized boolean canAcquire() 
+    {
+        long nowMicros = this.stopwatch.elapsed(TimeUnit.MICROSECONDS);
+        return delegate.canAcquire(nowMicros);
+    }
+
+    public synchronized long waitTimeMicros()
+    {
+        long nowMicros = this.stopwatch.elapsed(TimeUnit.MICROSECONDS);
+        return delegate.waitTimeMicros(nowMicros);
+    }
+
+    @VisibleForTesting
+    public synchronized long getNextFreeTicketMicros()
+    {
+        return delegate.nextFreeTicketMicros;
+    }
+
+    public synchronized double getStoredPermits()
+    {
+        return delegate.getStoredPermits();
+    }
+    
+    @Override
+    public String toString()
+    {
+        return String.format("Maximum requests/second is %.2f. %.2f stored permits available.", 
+                             getRate(), getStoredPermits());
+    }
+
+    @NotThreadSafe
+    private static class SmoothRateLimiter
+    {
+        private double storedPermits;
```

Comment:
For reviewers, carrying the concept of stored permits into this was partially a consequence of copy/paste from the Guava `SmoothBursty` rate limiter in the initial prototype. It's possible that we really don't need this burst capability at all, especially in a first cut of this.

...and of course, before this commits/merges, we should probably determine whether any licensing matter needs to decorate `SmoothRateLimiter`, although the Guava rate limiter is Apache 2.0 licensed.

CC @belliottsmith

Reply:
I'm not sure we really need burstiness, since we permit commands to execute immediately anyway, so that there's a kind of burst facility built in. If we implement our own (which we could, since it is quite simple), all I would say is that we might want to have some function of the maximum number of accumulated permits, i.e. a bound on the offset we are able to wait. This could be set to 1s, so that a period of quiescence can be carried forward. I'm not absolutely certain what licensing concerns there are, but I think we just need to include the original Guava license header. I'm sure there's some guidance somewhere.

Reply:
I think my plan at this point is just to remove the stored permits concept. When it comes to the number of accumulated permits, it seems like simply pausing connections will put a de-facto limit on how far into the future we might extend acquisition.

RE licensing, my plan is probably to throw the Guava license header on top of my inner class here.

Reply:
I tinkered a bit with removing the stored permits concept altogether, but in the end, even a very short burst window, say a tenth of a second, seems to allow for much tighter adherence to the specified rate. It's not terribly complicated, and we can always tweak the default I've assigned later. For now I'll keep it.